feat: add NeMo Guardrails component

jeffreyscarpenter opened the pull request:
Adds a new component for NeMo Guardrails for LLM Safety

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added support for NVIDIA NeMo Guardrails, enabling advanced guardrail options for large language model interactions, including self-checks, hallucination detection, content safety, topic control, and jailbreak detection.
  * Users can configure guardrails via multiselect options or custom YAML configurations.
  * Automatic fetching and selection of available self-check models from NVIDIA's API.

* **Bug Fixes**
  * Improved error handling and validation for required fields when enabling specific guardrails.

* **Tests**
  * Introduced comprehensive tests for the new NVIDIA NeMo Guardrails component to ensure reliability and correct behavior.

* **Chores**
  * Updated dependencies and version constraints to support the new NeMo Guardrails integration.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jeffreyscarpenter committed:
nemo guardrails component

jeffreyscarpenter committed:
integrating with backend services

jeffreyscarpenter committed:
content safety integration

jeffreyscarpenter committed:
override bot not respond

jeffreyscarpenter committed:
self check rails

jeffreyscarpenter committed:
self check model

jeffreyscarpenter committed:
self check model syntax

jeffreyscarpenter committed:
configuration change

jeffreyscarpenter committed:
update default jailbreak detection endpoint

jeffreyscarpenter committed:
linter automatic fixes

jeffreyscarpenter committed:
linter fixes

jeevic committed:
feat: endpoint basic examples add cache Increase interface speed (#7124)

* Update py_autofix.yml

* feat: endpoint basic_examples add cache

* Revert "Update py_autofix.yml"

This reverts commit 7aed45e98602f57c2d6d33532080bad30cf02507.

* resolve conflicts

* refactor: update response handling in read_basic_examples function

---------

Co-authored-by: tianzhipeng <tzpabc@gmail.com>
Co-authored-by: caojianwei-jk <caojianwei-jk@360shuke.com>

ogabrielluiz committed:
fix: Use async file operations for error log management in MCPStdioClient (#8012)

Cristhianzl committed:
fix: revert #7970 and add tags to tests (#8019)

ogabrielluiz committed:
chore: update CI workflow to check PyPI package version updates (#8025)

chore: update CI workflow to check for PyPI package version updates instead of nightly status

deon-sanchez committed:
fix: edit flow name settings not to be empty (#8020)

* Revert "feat: enhance flow settings with minimum length validation and default naming (#7970)"

This reverts commit 116bee825c043a8c11bcc86eb283950544c90537.

* feat: enhance flow settings with minimum length validation and default naming (#7970)

* feat: enhance flow settings with minimum length validation and default naming

* fix: remove console log from getFlowToAddToCanvas function

* test: add delay before filling flow name input to ensure stability

---------

Co-authored-by: cristhianzl <cristhian.lousa@gmail.com>

wbrennan899 committed:
feat: Twelve Labs Bundle (#7837)

* feat: add twelve labs components

* fix: fix uv.lock for twelve labs components

lucaseduoli committed:
feat: add bulk download and delete flows (#7849)

* update input to have h-fit

* Update McpServerTab text size

* Update Tools Component

* Update header text sizes

* Update list component to match design

* Update home page paddings

* Update home page to use ListComponent in both views

* Delete Grid

* Update skeleton to match design

* Remove old grid reference

* Implemented different border on checkbox

* Added selected flows

* Added selected flows action buttons

* Added flow selection on list component

* Added get download flows

* Added download and delete functions

* change download flows to download one flow directly

* implement shift selection

* Fix ctrl and meta behavior on selection

* remove selected flows if they dont exist

* added control just if its not mac

* Updated deletion modal

* Fixed delete confirmation modal taking up space in grid

* Fixed data-testids and success messages

* Added bulk actions test and fixed actionsMainPage

* added max width to home page

* [autofix.ci] apply automated fixes

---------

Co-authored-by: autofix-ci[bot] <114827586+autofix-ci[bot]@users.noreply.github.com>

mendonk committed:
docs: add webhook auth note (#8015)

* webhook-update

* remove-trailing-space-to-restart-ci

edwinjosechittilappilly committed:
feat: add default user agent in api request component  (#7631)

* Update api_request.py

* Update api_request.py

* Update api_request.py

* [autofix.ci] apply automated fixes

* Update src/backend/base/langflow/components/data/api_request.py

Co-authored-by: Gabriel Luiz Freitas Almeida <gabriel@langflow.org>

* Update src/backend/base/langflow/components/data/api_request.py

Co-authored-by: Gabriel Luiz Freitas Almeida <gabriel@langflow.org>

* Update api_request.py

* [autofix.ci] apply automated fixes

* fix format

* fix format issues

* [autofix.ci] apply automated fixes

* Template Update

---------

Co-authored-by: autofix-ci[bot] <114827586+autofix-ci[bot]@users.noreply.github.com>
Co-authored-by: Gabriel Luiz Freitas Almeida <gabriel@langflow.org>

lucaseduoli committed:
fix: added check to upload .md files correctly (#8017)

Added check to upload file type if file.type does not exist

phact committed:
fix: voice_mode break fix (#8014)

* 🔧 (use-get-messages-polling.ts): refactor enqueuePolling method to simplify logic and improve readability
🔧 (use-get-messages-polling.ts): refactor startNextPolling method to remove unnecessary code and improve efficiency
🔧 (use-get-messages-polling.ts): refactor removeFromQueue method to simplify and improve maintainability
🔧 (use-get-messages-polling.ts): refactor useGetMessagesPollingMutation to handle stopping polling and removing from queue
🔧 (use-get-transactions.ts): add early return to getTransactionsFn to handle empty id case
🔧 (use-get-voice-list.ts): refactor useGetVoiceList function to handle empty elevenlabsApiKey case and improve readability
🔧 (audio-settings-dialog.tsx): update useGetVoiceList call to pass elevenLabsApiKey as argument

* queues should hold payloads not strings

* ✨ (use-get-global-variables.ts): add setGlobalVariablesEntities function to store global variables entities for use in the application
✨ (audio-settings-dialog.tsx): introduce debounced function to update ElevenLabs API key in global variables entities
📝 (globalVariables.ts): add setGlobalVariablesEntities function to the global variables store for managing entities
📝 (index.ts): define GlobalVariable type for global variables entities in the store

* ✨ (frontend): add support for extracting flowId from URL query parameters to allow direct linking to specific flow logs.

---------

Co-authored-by: cristhianzl <cristhian.lousa@gmail.com>

lucaseduoli committed:
fix: json selection not visible (#8028)

* Fixed json selection not appearing and selection color

* Added h full to json viewer

lucaseduoli committed:
fix: add overflow to table dropdown selection, removed Multiple selector from Structured Output (#8027)

* Fixed not being able to edit options because of overflow in Table Node Cell Editor

* Updated structured output to not have Multiple selection since it's enabled by default

lucaseduoli committed:
feat: add bulk file actions (#7827)

* fixed styling

* Enabled header checkbox selection

* Changed styling of selection

* Implemented bulk downloading and deleting

* Added delete bulk hook

* Added download bulk hook

* Fix backend to send extension in download single file

* Fix hook to download single file directly

* Added header and selection handling

* Added delete confirmation

* [autofix.ci] apply automated fixes

* Fixed selection with shift

* Show disabled files

* Show disabled files as not clickable

* Changed color of icon when disabled

* Implemented pressed shift handling

* Fixed shift selection and disabled text selection when holding shift

* Created test for bulk selection on files modal

* add test of disabled components in file component

* Fixed files page test to include bulk editing test

* removed ring on focus visible

* Changed delete files having the right select

---------

Co-authored-by: autofix-ci[bot] <114827586+autofix-ci[bot]@users.noreply.github.com>

Cristhianzl committed:
fix: Enhance DB queries with async pagination and improve test reliability (#8040)

Co-authored-by: Lucas Oliveira <lucas.edu.oli@hotmail.com>
Co-authored-by: autofix-ci[bot] <114827586+autofix-ci[bot]@users.noreply.github.com>

mendonk committed:
docs: add bigquery integration (#8024)

* docs: add bigquery integration

* link-to-datasets

* cleanup

* remove-description-tags

Cristhianzl committed:
feat: Replace standard components with customizable versions and add Langflow Desktop support (#7904)

* ✨ (frontend): Add new customization components and hooks for the frontend application. This includes DashboardWrapperPage, McpServerTab, BaseModal, EmptyPageCommunity, FileCard, IOFileInput, IOModal, ProfilePictureChooserComponent, VoiceAssistant, App, router, CustomPreloadImages, useStartConversation, useStartRecording, CustomAudioWorkletProcessor.

* ✨ (frontend): Add new customization components and utilities for NodeStatus, curl code, JS API code, Python API code, and widget code.

* ✨ (frontend): Add new customization components and controllers for the frontend application.

* ✨ (frontend): Add new customization components and controllers for the frontend application.

* add customization folder desktop

* ✨ (AccountMenu): Introduce CustomAccountMenu component to customize the account menu in the app header
🔄 (ProfileIcon): Refactor ProfileIcon component to CustomProfileIcon for customization purposes
🔄 (langflow-counts): Refactor LangflowCounts component to CustomLangflowCounts for customization purposes

* 🔧 Remove unused files related to dialog, langflow, buildUtils, and reactflowUtils components

* ✨ (frontend): Add new images and components for the frontend UI
📝 (frontend): Update custom hooks to have more descriptive names for better clarity and organization

* 🔧 Remove package-lock.json as it is not needed and can be generated dynamically by npm or yarn when installing dependencies

* [autofix.ci] apply automated fixes

* 📝 (frontend): Add custom dialog content component without fixed positioning
🔧 (frontend): Update dialog class utility to improve dialog styling
🔧 (frontend): Update custom build utils for API endpoints
🔧 (frontend): Add custom get access token utility function
🔧 (frontend): Add custom get host protocol utility function
🔧 (frontend): Add custom pre-load image URL utility function
🔧 (frontend): Add custom react flow utility functions
🔧 (frontend): Add custom refresh latest version utility function
🔧 (frontend): Add custom routes for store pages
🔧 (frontend): Add custom routes for store pages components
🔧 (frontend): Update build utils for API endpoints
🔧 (frontend): Update react flow utility functions for downloading flows

* ✨ (custom-routes-store-pages.tsx): Rename customRoutesStorePages to CustomRoutesStorePages for consistency and clarity
✨ (custom-routes-store.tsx): Rename customRoutesStore to CustomRoutesStore for consistency and clarity
🔧 (routes.tsx): Remove unused import of StoreGuard to clean up the code and improve maintainability
🔧 (routes.tsx): Update imports of customRoutesStore and customRoutesStorePages to use named imports for consistency and clarity

* ✨ (API/index.ts): Add custom functions to fetch app versions and latest version from a custom source
🔧 (feature-flags.ts): Remove ENABLE_LANGFLOW_DESKTOP feature flag as it is no longer needed
📝 (custom-get-app-latest-version.ts): Add custom functions to fetch app versions and latest version from a custom source

* ♻️ (frontend): refactor API controller to export getAppVersions and getLatestVersion functions instead of calling custom functions directly

* ✨ (deploy-dropdown.tsx): Add customMcpOpen function to handle target attribute in CustomLink component
✨ (sideBarFolderButtons/index.tsx): Add CustomStoreButton component to render a custom store button if ENABLE_DATASTAX_LANGFLOW is false
📝 (custom-store-button.tsx): Create CustomStoreButton component to render a custom store button
📝 (custom-store-sidebar.tsx): Create CustomStoreSidebar component to define custom store sidebar items
📝 (custom-mcp-open.ts): Create customMcpOpen function to return "_blank" for target attribute
📝 (custom-refresh-latest-version.ts): Add useDarkStore import and use it to refresh latest version
📝 (SettingsPage/index.tsx): Import CustomStoreSidebar and use it to populate langflowItems
📝 (ProfilePictureChooserComponent/index.tsx): Import customPreLoadImageUrl and use it to preload images
📝 (use-preload-images.ts): Remove unused import of BASE_URL_API
📝 (use-preload-images.ts): Remove unused constant BASE_URL_API

* 🔧 (LangflowDesktopLogo.png): remove LangflowDesktopLogo.png as it is no longer needed in the project

* ✨ (custom-fetch-error-component.tsx): add CustomFetchErrorComponent to customize error handling in the frontend
🔧 (GenericErrorComponent/index.tsx): replace FetchErrorComponent with CustomFetchErrorComponent for more flexibility and customization in error handling

* ✨ (updater.ts): introduce new types Version and UpdaterStoreType for managing version information and updater modal state in the frontend customization module

---------

Co-authored-by: autofix-ci[bot] <114827586+autofix-ci[bot]@users.noreply.github.com>
Co-authored-by: Gabriel Luiz Freitas Almeida <gabriel@langflow.org>

schuellerf committed:
fix: remove double await in ChatOllamaComponent (#7992)

* Fix double await in ChatOllamaComponent

* ChatOllamaComponent: fix respect given URL self.base_url

* ChatOllamaComponent: implement support for "load_from_db" variables

* ChatOllamaComponent: fix RUF005 problem

* [autofix.ci] apply automated fixes

* conditionally await if variable is a coroutine

* Remove outdated comment regarding coroutine handling in ChatOllamaComponent

* ChatOllamaComponent: await coroutine result for JSON data retrieval

---------

Co-authored-by: autofix-ci[bot] <114827586+autofix-ci[bot]@users.noreply.github.com>
Co-authored-by: Gabriel Luiz Freitas Almeida <gabriel@langflow.org>

mendonk committed:
docs: mcp component name update (#8013)

* update-name-and-screenshots

* links

* update-astradb-screenshot

* sentence-case

* additional-linkfixes

thiau committed:
fix: Avoid unnecessary field resets in the API Request component (#7544)

* Add a logic to skip field reset if curl is not used

* Remove the else clause to avoid setting all other fields to advanced

* Fix return type

* Add a logic to avoid setting a field to advanced when use_curl is not defined

* Refactor else logic of _update_method_fields and _update_curl_mode

* Rollback _update_method_fields changes

* Return the current build config instead of an empty dict

---------

Co-authored-by: Edwin Jose <edwin.jose@datastax.com>
Co-authored-by: Gabriel Luiz Freitas Almeida <gabriel@langflow.org>

jeevic committed:
fix:  agent support  not use memory message (n_messages==0)  (#7800)

* Update py_autofix.yml

* feat: handle case when n_messages is zero in MemoryComponent

* agent memory support n_message zero

* Revert "Update py_autofix.yml"

This reverts commit 7aed45e98602f57c2d6d33532080bad30cf02507.

* fix: cast stored messages to Data type in retrieve_messages method

---------

Co-authored-by: tianzhipeng <tzpabc@gmail.com>

erichare committed:
fix: Hybrid search enabled as long as reranker is enabled (#8021)

* fix: Hybrid search enabled as long as reranker is enabled

* Back to original hybrid

* [autofix.ci] apply automated fixes

---------

Co-authored-by: autofix-ci[bot] <114827586+autofix-ci[bot]@users.noreply.github.com>

lucaseduoli committed:
fix: add scroll and max height on chosen files on File Component, fixed font weight on start page (#8045)

* Added max  height and scrolling to input file component

* Updated weight of titles

mendonk committed:
fix: api docs build failures caused by variables in descriptions (#8002)

* chat

* endpoints

phact committed:
fix: mcp enum fix (#8046)

default unknown ref name to string

mendonk committed:
docs: breaking changes in components feature (#8026)

* breaking-changes-feature

* dialog-not-popup

* dialog

* numbering

* include-individual-component-update

* fix-incorrect-information

* cleanup

lucaseduoli committed:
fix: remove connected edges when enabling tool mode for them to not reconnect later (#8047)

added tool mode cleaning on cleanEdges

lucaseduoli committed:
fix: made ctrl and shift state reset when switching windows to solve bulk editing issues (#8051)

* Fixed Ctrl Pressing when window is blurred

* Fixed Recent Files behavior with shift and ctrl

edwinjosechittilappilly committed:
fix: filter flow JSON from component JSON in MCP Servers (#8023)

filter for flows with ic_component as False

lucaseduoli committed:
fix: filter Update All components to not include edited ones, changed Node Toolbar to show Restore for edited componetns (#8050)

* Updated updateAllComponents to not include user edited components

* Changed GenericNode to display isOutdated to node toolbar if user edited

* Changed Restore color

---------

Co-authored-by: Edwin Jose <edwin.jose@datastax.com>

edwinjosechittilappilly committed:
fix: template update (#8057)

erichare committed:
fix: Properly escape new line characters in webhook (#8054)

Cristhianzl committed:
feat: centralize browser tab opening with customOpenNewTab utility (#8064)

* ✨ (NodeStatus): Add customOpenNewTab function to handle opening links in a new tab
🔧 (connectionComponent): Use customOpenNewTab function to open links in a new tab
🔧 (linkComponent): Use customOpenNewTab function to open links in a new tab
🔧 (custom-open-new-tab.ts): Create customOpenNewTab function to open links in a new tab
🔧 (new-modal): Use customOpenNewTab function to open LangflowButtonRedirectTarget link in a new tab
🔧 (nodeToolbarComponent): Remove unnecessary dependency openInNewTab from useEffect in NodeToolbarComponent

* ✨ (frontend): Add customOpenNewTab function to handle opening links in a new tab for better customization and control. Remove openInNewTab function to avoid redundancy and simplify code.

mendonk committed:
docs: remove duplicate entries from api examples page (#8065)

fix-duplicate-entries

mendonk committed:
docs: langflow docs review (#8032)

* quickstart

* message-history

* concepts-components

* data

* helpers

* logic-and-memories

* models

* agent-descriptions

* langsmith-more-env-vars

* use-kbd

* modify-descriptions

* Apply suggestions from code review

Co-authored-by: KimberlyFields <46325568+KimberlyFields@users.noreply.github.com>
Co-authored-by: Victor-w-Madeira <victor.wmadeira@gmail.com>

* remove-duplicate-periods

* remove-str-no-longer-used-in-ports

---------

Co-authored-by: KimberlyFields <46325568+KimberlyFields@users.noreply.github.com>
Co-authored-by: Victor-w-Madeira <victor.wmadeira@gmail.com>

mendonk committed:
docs: remove the loaders page and redirect to bundles (#8066)

remove-and-redirect-loaders-page-build-successful

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

ogabrielluiz committed:
fix: update docker build workflow to include 'nightly-main-all' release type for versioning (#8022)

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nemo_guardrails.py:
```diff
@@ -0,0 +1,571 @@
+            else:
+                logging.debug(f"  {header}: [REDACTED]")
+        logging.debug("Body:")
+        # Try to parse and redact sensitive info from body if it's JSON
```
I'd like to think we shouldn't need to do this here - maybe we can set up a shared logging config that does this.

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nemo_guardrails.py:
```diff
@@ -0,0 +1,571 @@
+
+    def get_models(self) -> list[str]:
+        """Get available models from NVIDIA API that are suitable for self-check tasks."""
+        # List of known good models for self-checking tasks
```
Are there specific models best for each rail?

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nemo_guardrails.py:
```diff
@@ -0,0 +1,571 @@
+            config = RailsConfig.from_content(yaml_content=yaml_content, colang_content=colang_content)
+
+            try:
+                config.model_validate(config)
```
The model should already be validating on initial construction, no need to do this again I believe.

jeffreyscarpenter replied:
verified that is correct, thanks, will remove

jeffreyscarpenter committed:
check model URLs are set for selected rails

jeffreyscarpenter committed:
nemo guardrails component unit tests

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

jeffreyscarpenter committed:
review feedback

jeffreyscarpenter committed:
Merge remote-tracking branch 'upstream/main' into feat/nemo-guardrails

jeffreyscarpenter committed:
Merge upstream/main into feat/nemo-guardrails: resolve NVIDIA component conflict and update imports

jeffreyscarpenter committed:
optional dependency guardrails 0.14.1

jeffreyscarpenter committed:
fix: linting issues for nest_asyncio workaround and dummy main model in nemo guardrails

- Use %%s logging instead of f-strings
- Assign exception messages to variables before raising
- Shorten long lines
- Remove unused arguments
- Fix multiline string syntax and line length

jeffreyscarpenter committed:
Clean up NVIDIANeMoGuardrailsComponent: remove unused imports, HTTP logging code, timer, and simplify build_model method

jeffreyscarpenter committed:
fix: resolve nemoguardrails dependency conflicts

- Make nemoguardrails truly optional by wrapping imports in try/except
- Add dependency overrides to resolve fastembed/huggingface-hub conflicts
- Pin compatible versions: fastembed==0.6.0, numpy>=1.26.0,<2.0.0
- Restore Python 3.13+ support while maintaining nemoguardrails compatibility

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

jeffreyscarpenter committed:
Fix: Ensure self-check model dropdown refresh button fetches models in NeMo Guardrails component

jeffreyscarpenter committed:
Refactor/Enhance: Move NeMo Guardrails rails config validation to dedicated method and require self-check model selection when self-check rails are enabled. Lint fixes for exception handling and if statements.

jeffreyscarpenter committed:
remove non-working model for self check

jeffreyscarpenter committed:
Merge branch 'main' into feat/nemo-guardrails

jordanrfrazier commented:
Can you update the field such that the selected are placed in new lines rather than comma separated? 

<img width="380" height="139" alt="Screenshot 2025-07-17 at 1 36 02 PM" src="https://github.com/user-attachments/assets/009af377-e3e7-4489-b7b1-ec5a2307f505" />

jordanrfrazier commented on src/backend/base/langflow/components/nvidia/nemo_guardrails.py:
```diff
@@ -0,0 +1,499 @@
+
+        config = RailsConfig.from_content(yaml_content=yaml_content, colang_content=colang_content)
+
+        return RunnableRails(config=config, llm=self.llm, verbose=self.guardrails_verbose)
```
This isn't an instance of `LanguageModel`. That said, I assume this was done to follow the pattern of `LCModelComponents`. 

Does this component need to inherit from the `LCModelComponent`? It's not necessarily a model, though it does follow some of the same I/O behaviors. 

If you want to leave it as-is, I would recommend updating the `outputs` to the following (or, similar, to match your methods) in order to remove the `Language Model` output option. 
```
    outputs = [
        Output(display_name="Response", name="text_output", method="text_response"),
    ]
```

jeffreyscarpenter replied:
I created the wrapper class.

jeffreyscarpenter committed:
fix: improve NeMo Guardrails integration with Langflow

- Add GuardrailsLanguageModel wrapper to satisfy LanguageModel type requirements
- Implement custom text_response method to properly handle input_value and system_message
- Fix input format for RunnableRails (expects {'input': string} not {'messages': [...]})
- Initialize variables to prevent UnboundLocalError
- Improve tracing integration - guardrails now appear as sub-spans instead of separate traces
- Maintain proper callback attachment for monitoring and debugging

jeffreyscarpenter committed:
Merge upstream/main into feat/nemo-guardrails

jeffreyscarpenter committed:
fix: resolve ruff linting issues in NeMo Guardrails test files

- Add missing __init__.py file to nvidia test directory
- Fix import sorting in test_nemo_guardrails.py
- Add proper punctuation to docstrings
- Remove unused mock parameter
- Add match parameter to pytest.raises
- Add trailing newline to __init__.py file

edwinjosechittilappilly commented:
Is this an expected response while installation?
<img width="999" height="892" alt="Screenshot 2025-07-21 at 1 02 38 AM" src="https://github.com/user-attachments/assets/e8bab8be-545f-4f26-8101-d1346fd85df7" />

jeffreyscarpenter committed:
fix(nemo-guardrails): always extract plain text from 'output' field in response for chat output

jeffreyscarpenter committed:
feat: hide system message input by default in NeMo Guardrails component

jeffreyscarpenter committed:
feat: improve Rails input options order and description in NeMo Guardrails component

jeffreyscarpenter committed:
feat: implement dynamic visibility for NeMo Guardrails inputs based on rails selection

jeffreyscarpenter committed:
feat: implement dynamic visibility for NVIDIA NeMo Guardrails component

- Add sophisticated dynamic visibility logic for model configuration and prompt inputs
- Group inputs logically by rail category (Content Safety, Topic Control, Jailbreak Detection, Self Check)
- Fix advanced=True fields to advanced=False for proper dynamic visibility
- Implement granular control where each prompt shows only when its specific rail is selected
- Organize input order to match rail dropdown order for consistent UX
- Add proper rail-to-input mapping for all configuration options

jeffreyscarpenter committed:
refactor: separate API keys for each rail type in NVIDIA NeMo Guardrails component

jeffreyscarpenter committed:
fix: set correct initial visibility for NVIDIA NeMo Guardrails inputs based on default rails selection

jeffreyscarpenter committed:
Merge remote-tracking branch 'upstream/main' into feat/nemo-guardrails

jeffreyscarpenter committed:
feat: add comprehensive optional dependency testing to CI

- Fix nemoguardrails test to handle optional dependency pattern
- Add new optional_dependencies_test.yml workflow for testing all optional deps
- Update main CI workflow to include optional dependency testing
- Follows established pattern of modular, reusable workflows

This ensures optional dependencies (nemoguardrails, audio, docling, etc.)
are properly tested when their dependencies are available, while maintaining
fast CI for core functionality.

jeffreyscarpenter committed:
refactor: improve optional dependency handling in nvidia components

- Refactor __init__.py to use cleaner, more scalable pattern
- Remove unnecessary NEMOGUARDRAILS_AVAILABLE and NV_INGEST_AVAILABLE variables
- Use += instead of append() for better type checker compatibility
- Maintain conditional imports for both nemoguardrails and nv-ingest
- Simplify exception handling with pass statements

This makes the code more maintainable and easier to extend with new optional components.

jeffreyscarpenter committed:
fix: move nemoguardrails imports to method level for optional dependency support

- Move RailsConfig and RunnableRails imports from module level to build_model method
- Add proper ImportError handling with helpful error message
- Update test to mock imports at the correct location
- Fixes component loading failure when nemoguardrails is not installed

This allows the component to be imported and loaded even when the optional
nemoguardrails dependency is not available, while still providing proper
functionality when it is installed.
